saleor/graphql/core/connection.py

```diff
@@ -1,7 +1,168 @@
 from graphene import Field, List, NonNull, ObjectType, String
 from graphene.relay.connection import Connection
 from graphene_django_optimizer.types import OptimizedDjangoObjectType
+from graphql_relay.connection.connectiontypes import Edge, PageInfo
+from graphql_relay.utils import base64, unbase64
+
+from ..core.enums import OrderDirection
+
+ConnectionArguments = Dict[str, Any]
+
+
+def to_global_cursor(values):
+    if not isinstance(values, List):
+        values = [str(values)]
+    else:
+        values = [str(value) for value in values]
+    return base64(":".join(values))
+
+
+def from_global_cursor(cursor) -> typing.List[str]:
+    values = unbase64(cursor)
+    values = values.split(":")
+    if isinstance(values, list):
+        return values
+    return [values]
+
+
+def connection_args_validation(args):
+    first = args.get("first")
+    last = args.get("last")
+
+    if isinstance(first, int) and first < 0 or first and not isinstance(first, int):
+        raise ValueError("Argument 'first' must be a non-negative integer.")
+    if isinstance(last, int) and last < 0 or last and not isinstance(last, int):
+        raise ValueError("Argument 'last' must be a non-negative integer.")
+    if first and last:
+        raise ValueError("Argument 'last' cannot be combined with 'first'.")
+    if first and args.get("before"):
+        raise ValueError("Argument 'first' cannot be combined with 'before'.")
+    if last and args.get("after"):
+        raise ValueError("Argument 'last' cannot be combined with 'after'.")
+
+
+def connection_from_queryset_slice(
+    qs: QuerySet,
+    args: ConnectionArguments = None,
+    connection_type: Any = Connection,
+    edge_type: Any = Edge,
+    page_info_type: Any = PageInfo,
+) -> Connection:
+    """Create a connection object from a QuerySet."""
+    args = args or {}
+    before = args.get("before")
+    after = args.get("after")
+    first = args.get("first")
+    last = args.get("last")
+    connection_args_validation(args)
+
+    cursor_after = from_global_cursor(after) if after else None
+    cursor_before = from_global_cursor(before) if before else None
+    cursor = cursor_after or cursor_before
+    cursor_offset = 1 if cursor else 0
+
+    sort_by = args.get("sort_by", {})
+    sorting_direction = sort_by.get("direction", "")
+    sorting_field = sort_by.get("field", "")
+
+    query_annotations = dict(qs.query.annotations) or {}
+    query_model = qs.query.model
+    if query_model:
+        model_fields = [field.name for field in query_model._meta.get_fields()]
+        default_ordering = query_model._meta.ordering
+        default_ordering = list(default_ordering) if default_ordering else []
+    else:
+        model_fields = default_ordering = []
+
+    if "sorting_value" in query_annotations:
+        order_expression = [f"{sorting_direction}sorting_value"] + default_ordering
+    elif sorting_field in query_annotations or sorting_field in model_fields:
```

Comment:
Why do we need this case? When is `sorting_value` not possible to have?

Reply:
I don't want to force everyone to add `.annotate(sorting_value=F("pk"))` everywhere
if you don't need to sort you result don't add it
every model should have default ordering set in meta class
this way we only need `sorting_value` in custom sort

Reply:
I would agree if the sorting was on PK but this code allows any field to be used as the sorting field without annotating the queryset. I don't think this function should care about `field` at all. This would save us from having to construct keyword arguments with `f""` expressions which I find very hard to read.